core/src/dpdk/port.rs

```diff
@@ -61,56 +62,68 @@ impl PortQueue {
     #[allow(clippy::trivially_copy_pass_by_ref)]
     pub fn receive(&self) -> Vec<Mbuf> {
         const RX_BURST_MAX: usize = 32;
-        let mut packets = Vec::with_capacity(RX_BURST_MAX);
+        let mut ptrs = Vec::with_capacity(RX_BURST_MAX);
 
         let len = unsafe {
             ffi::_rte_eth_rx_burst(
                 self.port_id.0,
                 self.rxq_index.0,
-                packets.as_mut_ptr(),
+                ptrs.as_mut_ptr(),
                 RX_BURST_MAX as u16,
             )
         };
 
-        unsafe {
-            packets.set_len(len as usize);
-        }
+        let mbufs = unsafe {
+            // does a no-copy conversion to avoid extra allocation.
+            Vec::from_raw_parts(ptrs.as_mut_ptr() as *mut Mbuf, len as usize, RX_BURST_MAX)
+        };
 
-        packets
-            .into_iter()
-            // should not have null pointers from rx burst
-            .map(|ptr| unsafe { ptr::NonNull::new_unchecked(ptr).into() })
-            .collect::<Vec<_>>()
+        mem::forget(ptrs);
+        mbufs
     }
 
     /// Sends the packets to the transmit queue.
     #[allow(clippy::trivially_copy_pass_by_ref)]
-    pub fn transmit(&self, packets: Vec<Mbuf>) {
-        let mut packets = packets.into_iter().map(Mbuf::into_ptr).collect::<Vec<_>>();
-
-        let mut to_send = packets.len() as u16;
-        while to_send > 0 {
+    pub fn transmit(&self, mut packets: Vec<Mbuf>) {
+        loop {
+            let to_send = packets.len() as u16;
             let sent = unsafe {
                 ffi::_rte_eth_tx_burst(
                     self.port_id.0,
                     self.txq_index.0,
-                    packets.as_mut_ptr(),
+                    packets.as_mut_ptr() as *mut *mut ffi::rte_mbuf,
```

Comment:
minor: i'd put a comment here as to why we use the **mut *mut* for the array of packets to mbufs, as it confused me a bit at first.